### ISSUE TITLE
Support for macOS Catalina?

### ISSUE THREAD

**Issue by MastaBaba** [stepmania/stepmania#1914]
The latest (beta) of Stepmania is not supported by macOS Catalina, the latest version of macOS that came out this week.

What's the expectation for making the two work together nicely?

**Comment by quietly-turning** [comment]
For the time being, it is necessary to [build your own SM5 from source](https://github.com/stepmania/stepmania/wiki/Compiling-StepMania#macos) for use with macOS 10.15.

I am not suggesting that this is an acceptable long-term solution for the StepMania project, but it is necessary until a new official release can be organized and put out.  It was my intention to open an issue here on GitHub detailing my thoughts on this topic some time tomorrow.

**Comment by quietly-turning** [comment]
> It was my intention to open an issue here on GitHub detailing my thoughts on this topic some time tomorrow.

This did not happen.  Sorry about that.  I'll try again with the next 24 hours I have available to me.

**Comment by MastaBaba** [comment]
How about today?

**Comment by quietly-turning** [comment]
I don't think I have it in me right now; winter came first.  I am sorry about this.

Please ping me again in May 2020 if this is not resolved by then and I'll try again at that time.

**Comment by Horsey-** [comment]
PM me and I can send you my pre-compiled SM5_1 from a few months ago. It’s very out of date, but it works. I won’t have access to OSX until the new MacBook Pro 16” is released (I returned my 15” to Costco when I heard the new one was coming out soon lol)

**Comment by Schwancky** [comment]
email sent Horsey

**Comment by Schwancky** [comment]
message sent horsey

**Comment by coderjo** [comment]
So, does this just need a 64-bit build, or does it need to be signed with a developer id and notarized by Apple?

**Comment by Schwancky** [comment]
I think it just needs 64 bit support.  32 no longer supported

**Comment by Horsey-** [comment]
OS X apps don’t need to be signed by Apple as long as you allow apps from outside sources (somewhere in settings, maybe under privacy?) again, I don’t have a Mac right now to take a screenshot of where the setting is

**Comment by ghost** [comment]
The Xcode project isn't building for me, exiting below:

Am I missing something?

```
yasm/nasm not found or too old. Use --disable-yasm for a crippled build.

If you think configure made a mistake, make sure you are using the latest
version from Git.  If the latest version fails, report the problem to the
ffmpeg-user@ffmpeg.org mailing list or IRC #ffmpeg on irc.freenode.net.
Include the log file "config.log" produced by configure as this will help
solving the problem.
make: *** [/Users/elliottmalone/stepmania/Build/ffmpeg-prefix/src/ffmpeg-stamp/Debug/ffmpeg-configure] Error 1
Command /bin/sh failed with exit code 2
```

**Comment by Horsey-** [comment]
> message sent horsey

@Schwancky tried my 64 bit compiled binary from september-ish and said it didn't work; I'll try again when the MBP 16 is released and I can get my hands on one.

**Comment by Schwancky** [comment]
thank ya

**Comment by halo** [comment]
@legareal I got past the `yasm` error by installing it with homebrew first. Then I ran `which yasm` to see where it was installed. I then edited the file `Build/CMakeCache.txt` and fixed the path here:

```cmake
# I changed this:
FFMPEG_YASM_EXECUTABLE:FILEPATH=FFMPEG_YASM_EXECUTABLE-NOTFOUND

# To something like this:
FFMPEG_YASM_EXECUTABLE:FILEPATH=/whatever/path/you/have/to/bin/yasm
```

Then I could build the project in Xcode (after having followed the instructions to change the Debug build to Release).  I moved the resulting Stepmania.app to `/Applications`.

After starting Stepmania, I was haunted with popup-errors saying things like

> Metric "DancingCamera::AmbientColor" is missing.  Correct this and click Retry, or Cancel to break.

That was because my `/Applications/Stepmania/Themes/_fallback` directory was outdated. I simply had to copy that directory from this git repository to where I had mine and now everything works on Catalina. The resulting app is 25 MB.

@dguzek I don't know how you did it, but I anticipated a **much** harder ride :) The cmake+Xcode combo worked surprisingly smoothly. Thank you for making it easy for me :)

**EDIT** More people are going to be bitten by the yasm error, I suggest improving that in the cmake flow. The "help wanted" label might also be removed from this issue, since there is no coding involved in this issue, I believe.

**Comment by quietly-turning** [comment]
Thanks for detailing your work here, @halo.  It is appreciated so that others developers may continue to build 64bit macOS applications using Xcode.

I very much still anticipate that the problem of organizing a formal SM5.1 release will be difficult, and I do not have that in me right now.  I am sorry about this.

**Comment by hayatae** [comment]
I followed the build instructions and successfully built an executable, but when I try to run, it crashes immediately:

```
StepMania5.1-git-6a4551f
Compiled 20191109 @ 22:28:35 (build 6a4551f)
Log starting 2019-11-10 12:25:16
Loading window: macosx
Model: MacBookPro10,1 (8/8)
Clock speed 2.30 GHz
Mac OS X 10.15.1
Memory: 8.00 GB
/////////////////////////////////////////
WARNING: Unable to determine system language. Using English.
/////////////////////////////////////////
2019-11-10 12:25:17.022942-0500 StepMania[31768:348241] [plugin] AddInstanceForFactory: No factory registered for id <CFUUID 0x600001745b60> F8BB1C28-BAE8-11D6-9C31-00039315CD46
2019-11-10 12:25:17.046957-0500 StepMania[31768:348241]  HALC_ShellDriverPlugIn::Open: Can't get a pointer to the Open routine
2019-11-10 12:25:17.048473-0500 StepMania[31768:348241]  HALC_ShellDriverPlugIn::Open: Can't get a pointer to the Open routine
Sound driver: AudioUnit
Lights driver: SystemMessage
/////////////////////////////////////////
WARNING: Unable to determine system language. Using English.
/////////////////////////////////////////
Video renderers: 'opengl'
2019-11-10 12:25:17.190104-0500 StepMania[31768:348225] Metal API Validation Enabled
Fatal: crash from within the crash signal handler
Program ended with exit code: 1
```

**Comment by heshuimu** [comment]
I have opened a PR #1933 as a temporary fix to resolve crashes on launch seen on Catalina. 
CC @hayatae

**Comment by heshuimu** [comment]
It does seem that there is a considerable amount of legacy Apple API usage (especially OpenGL oof). 
Apple is not known to be generous on backwards compatibility, maybe it's time to rewrite the macOS foundation? Metal support?

**Comment by heshuimu** [comment]
Also, first seen in Catalina, user must grant StepMania "Input Monitoring" privilege in System Preferences > Security & Privacy, otherwise keyboard input would not work.

**Comment by ChronoAndross** [comment]
@heshuimu 

https://github.com/stepmania/stepmania/issues/1895

I remember I asked about Metal/Vulkan conversion a couple months ago for Stepmania here and I think this could be related to the crashes we are seeing here. I do know that while OSX doesn't officially support OpenGL anymore since they are trying to push folks to Metal (yuck, but that's the way it is), Apple devices are still able to run OpenGL safely under a Metal context window providing it uses OpenGL 3.2 or above.

I found this tutorial online of how to do this: https://developer.apple.com/documentation/metal/mixing_metal_and_opengl_rendering_in_a_view

Perhaps now is the time to collect information about this and make a decision pertaining to this.

As a OSX user for Stepmania, I appreciate the dedication to making this work on Catalina. Thanks!

**Comment by alr1995** [comment]
Hello,
They released StepMania 5.3 compatible with Catalina. 
Only problem now is we need a tutorial on how to install it on mac. I tried the same method for the older StepMania version but didn't work. Can anyone help?

Download links:  https://projectmoon.dance/index.php?id=2

**Comment by ChronoAndross** [comment]
@alr1995 How is this project not advertised on here? I've never heard of this mod until now.

Anyway, I think if you scroll to the top of this page you'll see that you have to build SM from source if you're planning to run it on OSX Catalina. It sounds like there's going to be some coordinated effort to create these builds, but I'm not sure where that's going at this point.

**Comment by alr1995** [comment]
@ChronoAndross thank u! Yeah, I don't know why its not on here yet, I found it through StepMania official website in their post "the future of stepmania 5.3" https://www.stepmania.com

But yeah I really love the game but guess I will have to wait until someone comes up with another way to make it work on Catalina coz I'm not a tech person so I can't do this stuff!

**Comment by heshuimu** [comment]
Since the macOS support is already being actively worked on, I'll close my PR.

**Comment by hayatae** [comment]
FYI, stepmania outfox 5.3.0-alpha 4.5 (04/04/2020) from https://projectmoon.dance/index.php?id=2 is working great on macOS Catalina (10.15.3).

All I had to do to get it to work was to download the .dmg file from the site above, extract it to a folder (I just put mine in a folder called "stepmania" in my user's home directory), and then run the following command to remove the quarantine setting from all files in that directory:
`xattr -dr com.apple.quarantine /Users/your-username/stepmania`

Basically still the same steps from this guide: https://github.com/stepmania/stepmania/wiki/Installing-on-macOS

**Comment by joeheyming** [comment]
I have an issue, I'm not sure you have solved, but I used to connect my dance mats with a ps2 controller to usb converter.  The driver doesn't work on mac anymore.  I think it is the same 32 bit / 64 bit issue.  

This thing is ancient
https://www.play-asia.com/dual-shooter-ems-usb2/13/701b5

Any ideas?  Are there alternative dance mats that are usb-c compatible?

**Comment by Tatsh** [comment]
I used to own one of those. Unfortunately I don't know where it went. But I probably ran into problems with it that made it not worth using.

Those are [known to be problematic](https://bugs.launchpad.net/bugs/106622) [even on Windows](https://www.stepmania.com/forums/input-adapters-and-controllers/show/282) so I would suggest getting a new adaptor. More than likely none are going to be USB-C compatible (right now) so you will need an adaptor for that too. Hopefully lag isn't introduced by that but if it does the global offset should fix it.

**Comment by travankor** [comment]
>Please ping me again in May 2020 if this is not resolved by then and I'll try again at that time.

@quietly-turning 

>I am not suggesting that this is an acceptable long-term solution for the StepMania project, but it is necessary until a new official release can be organized and put out. It was my intention to open an issue here on GitHub detailing my thoughts on this topic some time tomorrow.

**Comment by quietly-turning** [comment]
Thanks for the ping, @travankor.

My StepMania time/energy right now is 100% dedicated to improving documentation for future themers.  I think it's important for me to complete that work first before taking on new tasks.

Is support for macOS Catalina something you can help with?

**Comment by travankor** [comment]
@quietly-turning 

Just for clarification, I was more interested in the bit where you say you would provide more details regarding a "new official release can be organized and put out". I think version 5.3 already has better support, but I don't know what the release process looks like.

Apologies if I misunderstood something.

**Comment by quietly-turning** [comment]
Gotcha, okay.

> I don't know what the release process looks like.

There's not one in place that I know of, but I'm also not a project owner/in charge of SM, so take what I say with a grain of salt.  🙂

---

> I was more interested in the bit where you say you would provide more details regarding a "new official release can be organized and put out".

As of 6bea7639e8c184ca5d681d5d2f665c71eef64d44 on `5_1-new`, the only absolute showstopper I am aware of is that StepMania crashes immediately on macOS.  The work done in #1933 provides a working fix for this.  If it were to be merged, a release "could happen."  

There are myriad problems with SM that could be fixed/improved, but there always will be.  I think the bigger issue is organizing a new release, rather than any one technical detail.  

SM5.0.12 has [just shy of 1M downloads](https://somsubhra.com/github-release-stats/?username=stepmania&repository=stepmania), so I think it's important to plan a new release carefully.

Does it make sense to release now?  What features/problems should determine that?  Who should weigh in on it?  Who will organize the release notes and author social media announcements that non-technical audiences can digest?

These questions *can* be answered, and there *can* be an SM5.1.  

Personally, I'd love for there to be another formal release, but there is more work involved I can manage by myself.

**Comment by DroutSound** [comment]
@quietly-turning Hey there, with the release of MacOS monterey this year, is there any way support will come for it?

**Comment by quietly-turning** [comment]
Hi @DroutSound; thanks for asking about this.  Regarding macOS support — I don't know.

I [haven't had a consistent way to play StepMania since 2017](https://twitter.com/quietly_turning/status/1458613302058266627) and found it best for my mental health to not shape my hours around continuing to support it. 💛
